### DISCUSSION TITLE
Why class is used in API call implementation, and what are the benefits of it?

### DISCUSSION THREAD

**Discussion by SaltMango** [infinitered/ignite#2146] [Q&A] [answered]
Why class is used in API call implementation, and what are the benefits of it?

**Comment by jamonholmgren** [comment, marked as answer]
It’s on my list to revisit later, but the primary benefits are that it can hold an apisauce instance and organize the api calls accordingly. Additionally, it’s easier to mock and test. But I’m looking at some alternate patterns too.

**Reply by mrshahzeb7** [reply]
@jamonholmgren , I have a question.

In api.ts, we have a getEpisodes function in which we call the API with all the conditions, along with a try-catch block (which repeats with every call). 

Shouldn't we create something like getMethod, postMethod, deleteMethod, etc., so that we can have a bottleneck through which every call passes? 

Are there any drawbacks to this approach?

**Reply by jamonholmgren** [reply]
@mrshahzeb7 I think that's an interesting approach. Seems reasonable to me. Do you have a use case in mind?

I'm also interested in updating this to not use a class but rather just functions. I asked ChatGPT to give me an option and [it actually looks pretty good](https://chat.openai.com/share/9a0b10e8-0a4c-4156-931b-e2f9c1c2a7b9). What do you think of this new slightly different pattern?

**Reply by mrshahzeb7** [reply]
I shared the example I have in mind on git-gist It is not a working example just an idea
https://gist.github.com/mrshahzeb7/b4f3417f8678ccf471b22f1a875df7e3

1. Create Singleton instance(Class) of apiSauce (by doing this we can create different files based on backend controllers like user.ts, episode.ts, feed.ts  and tours.ts they all will utilize the same instance of apisauce)
Class will only be used for instance, for the rest we can use functions

2. Create separate functions for API methods (get, post, delete, put) but they all have to go through a single function, so we can use something like this It won't bloat our current file,  callApi(endpoints.users, methods.get)

3. Create a mapper function if you want to test API response and pass type to it if it is correct it will return the response otherwise will return bad data

4: This line should also be improved. it is repeated in every request Promise<
  { kind: "ok"; episodes: EpisodeSnapshotIn[] } | GeneralApiProblem
>
Create something like this ReturnResponse<EpisodeSnapshotIn[]> it will reduce our code a lot
It is just an idea I am not as good as you guys at using TypeScript, I learned mostly from Ignite boilerplate 😃
